Claim module - April 5 rebase (#72)
osmosis-labs/osmosis: x/claim/spec/README.md

# `claim`

## Abstract

This specifies the `claim` module of Osmosis project, provide commands for claimable amount query and claim airdrop.

> We apply real-time decay after `DurationUntilDecay` pass where monthly decay rate is `-10%` of inital airdrop amount.

Reviewer: Is the `-10%` a parameter or hard coded?

Reviewer: Also feels odd to me that we don't just linearly decay until 0%

Author: What kind of decay mechanism is better for this?

> When `DurationOfDecay` time pass, all coins are funded to community pool.

Reviewer: ```suggestion
When `DurationUntilDecay + DurationOfDecay` time pass, all coins are funded to community pool.
```

## Genesis State

### Accounts

All genesis accounts have `1 Osmo` for claim fee.

### Claimables

> Claimables are configured by genesis.

Reviewer: Can you define what a claimable is here?

Author: ok

### Airdrop Tools
There are tools to generate genesis from cosmos-hub snapshot.

#### Genesis generation

Generate genesis from cosmos-hub snapshot genesis and output snapshot of atom, osmo balance and percentage by address.
```sh
osmosisd export-airdrop-genesis uatom ../genesis.json --total-amount=100000000000000 --snapshot-output="../snapshot.json"
osmosisd export-airdrop-genesis uatom ../genesis.json --snapshot-output="../snapshot.json"
```

## Queries

> Query claimable amount of current time.

Reviewer: ```suggestion
Query claimable amount for a given address at the current time.
```

```sh
osmosisd query claim claimable $(osmosisd keys show -a validator --keyring-backend=test)
```

## Msgs

> Claim full airdrop amount from `claim` module.

Reviewer: ```suggestion
### (WIP) Actual claim commands will change

Claim full airdrop amount from `claim` module.
```

# `claim`

## Abstract

This specifies the `claim` module of Osmosis project, provide commands for claimable amount query and claim airdrop.
We apply real-time decay after `DurationUntilDecay` pass where monthly decay rate is `-10%` of inital airdrop amount.
When `DurationUntilDecay + DurationOfDecay` time passes, all unclaimed coins will be sent to the community pool.

## Genesis State

### Accounts

> All genesis accounts have `1 Osmo` for claim fee.

Reviewer: ```suggestion
All genesis accounts have `1 Osmo` for tx fees prior to claiming
```

```

### User withdrawn actions
A user can't withdraw twice for single action type, example `Vote`. To withdraw all, they need to do every action.
To do this, claim module manage withdrawn actions flag. It's using same strategy as user actions, but only count the actions that user has already withdrawn.
If Alice has done `DelegateStake` action, withdraw 25% of claimables, and then if she do `Vote` action, she is able to withdraw another 25% of claimables. Here, withdrawn actions flag is used to avoid double withdraw on `DelegateStake` action.

### Airdrop Tools
There are tools to generate genesis from cosmos-hub snapshot.

#### Genesis generation

Generate genesis from cosmos-hub snapshot genesis and output snapshot of atom, osmo balance and percentage by address.
```sh
osmosisd export-airdrop-genesis uatom ../genesis.json --total-amount=100000000000000 --snapshot-output="../snapshot.json"
osmosisd export-airdrop-genesis uatom ../genesis.json --snapshot-output="../snapshot.json"
```

## Queries

Query claimable amount for a given address at the current time.
```sh
osmosisd query claim claimable $(osmosisd keys show -a validator --keyring-backend=test)
```

> ## Msgs

Reviewer 2: Why do people have to manually send a claim msg?  Why don't we just transfer the coins during the hook itself (when the flag is set)?

Author: It would be very good to do it within hook, manual interaction is making codebase complex

Reviewer 2: Let's go ahead and do that.  Yeah, seems like it would get rid of a lot of the complexity.

Author: Yes, it's done, can you check again?
